### PR TITLE
LKE-12367 Smart layouts in the group nodes

### PR DESCRIPTION
**Pull request by w8r**
Added more options to subnodes layout to make it faster
LKE-12367

### CODE REVIEW

**Comment by aminerhanemi** [comment]
I tested the changes in this [viz](https://int-linkurious-enterprise-develop-stable.preview.k8s.dev.linkurious.net/workspace/721?key=finc0100) 
Running the layout on sub-nodes went on average from 1243.ms to 621ms

File: src/ogma/features/nodeGrouping.ts [hunk 1]
```diff
@@ -282,4 +331,71 @@ export class NodeGroupingTransformation {
       )}-${propertyValue}`
     );
   }
+
+  public _runCircularLayout({
```
Comment (aminerhanemi) [inline src/ogma/features/nodeGrouping.ts:334]: this is duplicate of circularLayout in ogmaTool,

File: src/ogma/features/nodeGrouping.ts [hunk 2]
```diff
@@ -147,12 +157,51 @@ export class NodeGroupingTransformation {
...
+    // 2 nodes
+    if (subNodes.size === 2) {
+      const radii = subNodes.getAttribute('radius').map(Number);
+      const positions = subNodes.getPosition();
+      const gap = Math.min(...radii);
+      await subNodes.setAttributes([
+        positions[0],
+        {x: positions[0].x + gap + radii[0] + radii[1], y: positions[0].y}
+      ]);
+      return;
```
Comment (aminerhanemi) [inline src/ogma/features/nodeGrouping.ts]: better move it to a method runTwoNodesLayout

File: src/ogma/features/nodeGrouping.ts [hunk 3]
```diff
@@ -147,12 +157,51 @@ export class NodeGroupingTransformation {
...
+      const positions = this._runCircularLayout({
+        radii: satellites.getAttribute('radius'),
+        cx: center.getAttribute('x'),
+        cy: center.getAttribute('y'),
+        clockwise: false,
+        distanceRatio: 5
+      });
+      const list = center.toList().concat(satellites);
+      await list.setAttributes([center.getPosition(), ...positions]);
+      return;
```
Comment (aminerhanemi) [inline src/ogma/features/nodeGrouping.ts:182]: same we can move to a separate method

File: src/ogma/features/nodeGrouping.ts [hunk 4]
```diff
@@ -147,12 +157,51 @@ export class NodeGroupingTransformation {
...
+    const degrees = subNodes.getDegree();
+    if (degrees.every((d) => d > 0 && d <= 2)) {
+      // straighten the chain
+      const sortedNodes = this._ogma.getNodes(this.topologicalSort(subNodes));
+      // we also need to sort the nodes so that they are following the chain
+      await this._ogma.layouts.grid({
+        nodes: sortedNodes,
+        // TODO: test that visually
+        colDistance: Math.max(...subNodes.getAttribute('radius').map(Number)) * 4,
+        rows: 1
```
Comment (aminerhanemi) [inline src/ogma/features/nodeGrouping.ts]: same

File: src/ogma/features/nodeGrouping.ts [hunk 5]
```diff
@@ -282,4 +331,71 @@ export class NodeGroupingTransformation {
...
+      const ry = r * Math.sin(angle);
+      return {
+        x: cx + rx,
+        y: cy + ry
+      };
+    });
+  }
+
+  public topologicalSort(nodes: NodeList) {
+    const nodesArray = nodes.toArray();
```
Comment (aminerhanemi) [inline src/ogma/features/nodeGrouping.ts]: we can remove it form here as it is moved to ogmaTool

File: src/ogma/features/nodeGrouping.ts [hunk 6]
```diff
@@ -147,12 +157,51 @@ export class NodeGroupingTransformation {
...
+      return;
+    }
+    // Chains: if al nodes have degree 1 or 2, place them in a line
+    const degrees = subNodes.getDegree();
+    if (degrees.every((d) => d > 0 && d <= 2)) {
+      // straighten the chain
+      const sortedNodes = this._ogma.getNodes(this.topologicalSort(subNodes));
+      // we also need to sort the nodes so that they are following the chain
+      await this._ogma.layouts.grid({
+        nodes: sortedNodes,
```
Comment (aminerhanemi) [inline src/ogma/features/nodeGrouping.ts]: in the grid layout, the position of nodes will be defined by their index in the list ?

**Comment by aminerhanemi** [comment]
@w8r  ping for the review

File: src/ogma/features/nodeGrouping.ts (hunk 6 above)
Comment (w8r) [reply src/ogma/features/nodeGrouping.ts]: yes

File: src/ogma/features/nodeGrouping.ts (hunk 1 above)
Comment (w8r) [reply src/ogma/features/nodeGrouping.ts:334]: @aminerhanemi I didn't know where is the correct place to put it. What's the best way?

Comment (w8r) [reply src/ogma/features/nodeGrouping.ts:334]: left it in the NodeGrouping file

**Review by aminerhanemi — APPROVED** [review]

**Comment by w8r** [comment]
🔥

**Pull request merged**